### ISSUE TITLE
Unhandled Payment Failure (Stripe) and cart inconsistencies 

### ISSUE THREAD

**Issue by nalimhahs** [mirumee/saleor-storefront#848]
### What I'm trying to achieve
Complete a purchase.

### Steps to reproduce the problem
1. Add any product to cart and checkout from cart.
2. Login as registered user in the page that follows.
3. Select an address, choose shipping and in the payment page use stripe and enter a card which requires 3D secure authentication. I used 4000000000003220 here.
4. Proceed to review and place order. (No payment auth page popped up)
5. Order Successful Page appears but WITHOUT an order number. 
6. Trying to view the order details results in a 404 (With saleor core throwing a `ValueError: badly formed hexadecimal UUID string` error)
7. Opening the order history page shows no such order existing.

This is bug 1. Continuing,

8. Opening the cart shows it's empty. Now add a different item to the cart.  
9. Proceed to checkout. Take note of the cart summary it should show the current added item. 
10. Select a shipping address and proceed to next step. The cart will now show the previous failed item instead of the new one. 
11. Use the same card to perform payment results in the same error as before.

Repeating steps 8-11 results in the first failed item being shown again and again. 

### What I expected to happen
Successful Order Placement

### Screenshots

![Screenshot from 2020-08-07 15-36-31](https://user-images.githubusercontent.com/42489509/89635520-ba5bf600-d8c4-11ea-9cb4-8f1f6420ef6a.png)
Adding cushion to cart.

![image (2)](https://user-images.githubusercontent.com/42489509/89635546-c5168b00-d8c4-11ea-99a4-144d805ec4e3.jpg)
Cushion being shown in cart summary.

![Screenshot from 2020-08-07 15-37-11](https://user-images.githubusercontent.com/42489509/89635590-d495d400-d8c4-11ea-9831-ba8234cf7381.png)
Cart summary now showing shoes i tried to purchase earlier.

![Screenshot from 2020-08-07 15-38-57](https://user-images.githubusercontent.com/42489509/89635634-e4151d00-d8c4-11ea-8e2d-c7e6c121b844.png)
This shows the missing order number.

**System information**
Operating system: Ubuntu 18.04
Browser: Chrome

Running Saleor Core(2.10.2), Saleor Storefront (2.10.4 and 2.10.0)

**Comment by igoris** [comment]
I got this issue while trying to add my custom payment gateway and initially thought that there's a bug in my code. However, this is what I found after some digging:

1. Method `capture_payment` in the plugin on the backend returns `GatewayResponse` with `action_required=True` for 3D secure
2. Method `perform_mutation` of the class `CheckoutComplete` checks the value of `action_required`:
2.1. if it's `False`, creates and returns an order (https://github.com/mirumee/saleor/blob/master/saleor/graphql/checkout/mutations.py#L865)
2.2. if it's `True`, returns `order=None` and `confirmation_needed=True` (https://github.com/mirumee/saleor/blob/master/saleor/graphql/checkout/mutations.py#L867)
3. The client code doesn't request the `confirmation_needed` field (https://github.com/mirumee/saleor-sdk/blob/master/src/mutations/checkout.ts#L197)
4. The client code also doesn't check that the order was actually created (https://github.com/mirumee/saleor-storefront/blob/master/src/%40next/pages/CheckoutPage/subpages/CheckoutReviewSubpage.tsx#L84)

So, in short, the case where additional step is required is not handled on the storefront side at all.

Hope this helps

**Comment by maarcingebala** [comment]
Currently, Storefront doesn't handle 3DS payments, but we're just adding Adyen support which does that. Once this is merged, it should be easier to build support for 3DS in Stripe based on the Adyen implementation.
